Mathlib/NumberTheory/NumberField/Completion.lean

```diff
@@ -0,0 +1,306 @@
+absolute value. This provides a systematic way of assigning and inferring instances of the semiring
+that also depend on an absolute value. In our application, relevant instances and the completion
+of a number field `K` are first defined at the level of `AbsoluteValue` by using the type synonym
+`WithAbs` of `K`, and then derived downstream for `InfinitePlace` (which is a subtype of
+`AbsoluteValue`).
+
+## Main definitions
+ - `WithAbs` : type synonym for a semiring which depends on an absolute value. This is
+  a function that takes an absolute value on a semiring and returns the semiring. We use this
+  to assign and infer instances on a semiring that depend on absolute values.
+ - `AbsoluteValue.completion` : the uniform space completion of a field `K` according to the
+  uniform structure defined by the specified real absolute value.
+ - `NumberField.InfinitePlace.completion` : the completion of a number field `K` at an infinite
+  place, obtained by completing `K` with respect to the absolute value associated to the infinite
+  place.
+ - `NumberField.InfinitePlace.Completion.extensionEmbedding` : the embedding `v.embedding : K →+* ℂ`
+  extended to `v.completion →+* ℂ`.
+ - `NumberField.InfinitePlace.Completion.extensionEmbedding_of_isReal` : if the infinite place `v`
+  is real, then this extends the embedding `v.embedding_of_isReal : K →+* ℝ` to
+  `v.completion →+* ℝ`.
+ - `NumberField.InfinitePlace.Completion.equiv_real_of_isReal` : the ring isomorphism
+  `v.completion ≃+* ℝ` when `v` is a real infinite place; the forward direction of this is
+  `extensionEmbedding_of_isReal`.
+ - `NumberField.InfinitePlace.Completion.equiv_complex_of_isComplex` : the ring isomorphism
+  `v.completion ≃+* ℂ` when `v` is a complex infinite place; the forward direction of this is
+  `extensionEmbedding`.
+
+## Main results
+ - `NumberField.Completion.locallyCompactSpace` : the completion of a number field at
+  an infinite place is locally compact.
+ - `NumberField.Completion.isometry_extensionEmbedding` : the embedding `v.completion →+* ℂ` is
+  an isometry. See also `isometry_extensionEmbedding_of_isReal` for the corresponding result on
+  `v.completion →+* ℝ` when `v` is real.
+ - `NumberField.Completion.bijective_extensionEmbedding_of_isComplex` : the embedding
+  `v.completion →+* ℂ` is bijective when `v` is complex. See also
+  `bijective_extensionEmebdding_of_isReal` for the corresponding result for `v.completion →+* ℝ`
+  when `v` is real.
+
+## Tags
+number field, embeddings, infinite places, completion, absolute value
+-/
+noncomputable section
+
+/-- Type synonym for a semiring which depends on an absolute value. This is a function that takes
+an absolute value on a semiring and returns the semiring. We use this to assign and infer instances
+on a semiring that depend on absolute values. -/
+@[nolint unusedArguments]
+def WithAbs {R S : Type*} [Semiring R] [OrderedSemiring S] :
+    AbsoluteValue R S → Type _ := fun _ => R
```

Comment:
If we go for this then I don't think it should be in a file called `NumberTheory.NumberField.Completion`, it should probably exist in a new file or at least somewhere which isn't about number fields.

I'm not convinced about the need for this though; I understand the argument, but I'm wondering whether you can just use `abbrev completion := @UniformSpace.Completion K <| letI : NormedField K := v.toNormedField; inferInstance` (and toss in a few more `  letI : NormedField K := v.toNormedField;
` when necessary) to get around this rather than opening up the debate about adding a new definition. Maybe worth having a discussion on Zulip if you can't see my suggestion working. I can see the argument for a new declaration but some people might want to see more than one use case before agreeing.

Reply:
I agree that it should probably be in its own file, as was done for [`WithLp`](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Analysis/Normed/Lp/WithLp.html).

Reply:
You could move this declaration to a shortish file WithAbs.lean, put it somewhere appropriate, make some basic API for it e.g. it's a normed ring, it's a normed field if K is a field etc -- just get it out of this file and then hopefully this PR will be ready to be merged. You've convinced me that this approach is worth trying.

Reply:
Have now added `WithAbs.lean` with a small API modelled on `WithLp.lean`